x/concentrated-liquidity/pool.go

```diff
@@ -76,127 +95,177 @@ func (p Pool) SpotPrice(ctx sdk.Context, baseAssetDenom string, quoteAssetDenom
+	}
+	sqrtPCurTick := curSqrtPrice
+	sqrtPUpperTick, err := maxPrice.ApproxSqrt()
+	if err != nil {
+		return sdk.Coin{}, sdk.Coin{}, fmt.Errorf("issue calculating square root of maxPrice")
+	}
 
-	// TODO: Roman change out with query to pool to get this info
+	// TODO: How do we remove/generalize this? I am stumped.
 	amountETH := int64(1000000)
 	amountUSDC := int64(5000000000)
 
+	// find liquidity of assetA and assetB
 	liq0 := liquidity0(amountETH, sqrtPCurTick, sqrtPUpperTick)
 	liq1 := liquidity1(amountUSDC, sqrtPCurTick, sqrtPLowerTick)
 
+	// utilize the smaller liquidity between assetA and assetB when performing the swap calculation
 	liq := sdk.MinDec(liq0, liq1)
-	if tokenIn.Denom == asset1 {
-		priceDiff := tokenAmountInAfterFee.Quo(liq)
-		priceNext := sqrtPCurTick.Add(priceDiff)
-		// new amount in, will be needed later
-		//amountIn = calcAmount1(liq, priceNext, sqrtPCurTick)
-		amountOut := calcAmount0Delta(liq, priceNext, sqrtPCurTick)
-		return sdk.NewCoin(tokenOutDenom, amountOut.TruncateInt()), nil
-	} else if tokenIn.Denom == asset0 {
-		priceNextTop := liq.Mul(sqrtPCurTick)
-		priceNextBot := liq.Add(tokenAmountInAfterFee.Mul(sqrtPCurTick))
-		priceNext := priceNextTop.Quo(priceNextBot)
-		// new amount in, will be needed later
-		//amountIn = calcAmount1(liq, priceNext, sqrtPCurTick)
-		amountOut := calcAmount1Delta(liq, priceNext, sqrtPCurTick)
-		return sdk.NewCoin(tokenOutDenom, amountOut.TruncateInt()), nil
-	}
-	return sdk.Coin{}, fmt.Errorf("tokenIn (%s) does not match any asset in pool", tokenIn.Denom)
+
+	swapState := SwapState{
+		amountSpecifiedRemaining: tokenAmountInAfterFee,
+		amountCalculated:         sdk.ZeroDec(),
+		sqrtPrice:                sqrtPCurTick,
+		tick:                     priceToTick(sqrtPCurTick.Power(2)),
+	}
+	// TODO: This should be GT 0 but some instances have very small remainder
+	// need to look into fixing this
+	for swapState.amountSpecifiedRemaining.GT(sdk.NewDecWithPrec(1, 6)) {
+		stepState := StepState{}
+		stepState.sqrtPriceStart = swapState.sqrtPrice
+		lte := tokenIn.Denom == asset1
+		nextTick, _ := k.NextInitializedTick(ctx, poolId, swapState.tick.Int64(), lte)
+		stepState.nextTick = sdk.NewInt(nextTick)
+		nextSqrtPrice, _ := k.tickToPrice(stepState.nextTick).ApproxSqrt()
+		stepState.sqrtPriceNext = nextSqrtPrice
+
+		swapState.sqrtPrice, stepState.amountIn, stepState.amountOut = computeSwapStep(
+			swapState.sqrtPrice,
+			stepState.sqrtPriceNext,
+			liq,
+			swapState.amountSpecifiedRemaining,
+			lte,
+		)
+		swapState.amountSpecifiedRemaining = swapState.amountSpecifiedRemaining.Sub(stepState.amountIn)
+		swapState.amountCalculated = swapState.amountCalculated.Add(stepState.amountOut)
+		swapState.tick = priceToTick(swapState.sqrtPrice.Power(2))
+	}
+	newTokenIn.Amount = tokenIn.Amount.Sub(swapState.amountSpecifiedRemaining.RoundInt())
+	return sdk.NewCoin(tokenIn.Denom, newTokenIn.Amount), sdk.NewCoin(tokenOutDenom, swapState.amountCalculated.RoundInt()), nil
```

Comment:
I don't think we should be r using bankers rounding anywhere here. It should either be rounding up or down.

For example, for tokenIn, I would assume we should round up so that we don't undercharge user.

Reply:
changed the logic to round up and down as requested